### PR TITLE
Handle out of order repair notifications

### PR DESCRIPTION
**Pull request by adejanovski**
In the case repair sessions end in a very short time (within a few 10s of milliseconds) we can get out of order notifications from Cassandra :
Instead of `START -> SUCCESS -> COMPLETE`, we get `START -> COMPLETE -> SUCCESS`
This commit handles this case by not signaling the wait condition of the segment unless we have both a `SUCCESS` or an `ERROR` notification plus a `COMPLETE` one.

### CODE REVIEW

**Comment by michaelsembwever** [comment]
Wanna rebase this off master @adejanovski ?

File: src/server/src/test/java/io/cassandrareaper/service/SegmentRunnerTest.java [hunk 1]
```diff
@@ -475,6 +475,145 @@ protected JmxProxy connect(
     assertEquals(2, storage.getRepairSegment(runId, segmentId).get().getFailCount());
   }
 
+  @Test
+  public void outOfOrderTest() throws InterruptedException, ReaperException, ExecutionException {
```
Comment (michaelsembwever) [inline src/server/src/test/java/io/cassandrareaper/service/SegmentRunnerTest.java]: i have verified that this test fails without the patch applied. 👍

File: src/server/src/main/java/io/cassandrareaper/service/SegmentRunner.java [hunk 2]
```diff
@@ -124,6 +127,9 @@
     this.repairUnit = repairUnit;
     this.repairRunner = repairRunner;
     this.segmentFailed = new AtomicBoolean(false);
+    this.gotCompleteNotification = new AtomicBoolean(false);
+    this.gotFailedNotification = new AtomicBoolean(false);
+    this.gotSuccessNotification = new AtomicBoolean(false);
```
Comment (michaelsembwever) [inline src/server/src/main/java/io/cassandrareaper/service/SegmentRunner.java]: can we use the following shorter names for these plesae?
{{completeNotified}}, {{failedNotified}}, {{successNotified}}

File: src/server/src/main/java/io/cassandrareaper/service/SegmentRunner.java [hunk 3]
```diff
@@ -96,6 +96,9 @@
   private int commandId;
   private final AtomicBoolean segmentFailed;
   private final UUID leaderElectionId;
+  private final AtomicBoolean gotFailedNotification;
+  private final AtomicBoolean gotSuccessNotification;
+  private final AtomicBoolean gotCompleteNotification;
```
Comment (michaelsembwever) [inline src/server/src/main/java/io/cassandrareaper/service/SegmentRunner.java]: can we use just two AtomicBooleans, and use the following shorter names for these please?
`completeNotified`, `failedOrSuccessNotified`

File: src/server/src/main/java/io/cassandrareaper/service/SegmentRunner.java (hunk 2 above)
Comment (michaelsembwever) [inline src/server/src/main/java/io/cassandrareaper/service/SegmentRunner.java]: can we remove these 3 lines please, they are not needed (the fields can be initialised on lines 99-101)

File: src/server/src/main/java/io/cassandrareaper/service/SegmentRunner.java [hunk 4]
```diff
@@ -808,6 +814,14 @@ private boolean handleJmxNotificationForCassandra22(
                     .withEndTime(DateTime.now())
                     .withId(segmentId)
                     .build());
+
+            gotSuccessNotification.set(Boolean.TRUE);
```
Comment (michaelsembwever) [inline src/server/src/main/java/io/cassandrareaper/service/SegmentRunner.java]: autobox?
ie `completeNotified.set(true);`

File: src/server/src/main/java/io/cassandrareaper/service/RepairRunner.java [hunk 5]
```diff
@@ -149,6 +149,11 @@ static int getPossibleParallelRepairsCount(
       throw new ReaperException(msg);
     }
 
+    LOG.debug(
+        "Possible parallel repairs : {}",
+        Math.min(
+            ranges.size() / ranges.values().iterator().next().size(),
+            Math.max(1, hostsInRing.keySet().size() / ranges.values().iterator().next().size())));
```
Comment (michaelsembwever) [inline src/server/src/main/java/io/cassandrareaper/service/RepairRunner.java:156]: maybe wrap this debug statement inside `if (LOG.isDebugEnabled()) {…}`

File: src/server/src/main/java/io/cassandrareaper/service/SegmentRunner.java [hunk 6]
```diff
@@ -894,6 +920,14 @@ private boolean handleJmxNotificationForCassandra21(
                     .withId(segmentId)
                     .build());
 
+            // Since we can get out of order notifications,
+            // we need to exit if we already got the COMPLETE notification.
+            gotSuccessNotification.set(Boolean.TRUE);
+            if (gotCompleteNotification.get()) {
```
Comment (michaelsembwever) [inline src/server/src/main/java/io/cassandrareaper/service/SegmentRunner.java]: this will never be true

File: src/server/src/main/java/io/cassandrareaper/service/SegmentRunner.java [hunk 7]
```diff
@@ -903,19 +937,33 @@ private boolean handleJmxNotificationForCassandra21(
...
-        LOG.warn("repair session failed for segment with id '{}' and repair number '{}'", segmentId, repairNumber);
+        LOG.warn(
+            "repair session failed for segment with id '{}' and repair number '{}'",
+            segmentId,
+            repairNumber);
         failOutsideSynchronizedBlock = true;
+        // Since we can get out of order notifications,
+        // we need to exit if we already got the COMPLETE notification.
+        gotFailedNotification.set(Boolean.TRUE);
+        if (gotCompleteNotification.get()) {
```
Comment (michaelsembwever) [inline src/server/src/main/java/io/cassandrareaper/service/SegmentRunner.java]: this will never be true

File: src/server/src/main/java/io/cassandrareaper/service/SegmentRunner.java [hunk 8]
```diff
@@ -903,19 +937,33 @@ private boolean handleJmxNotificationForCassandra21(
...
 
       case SESSION_FAILED:
-        LOG.warn("repair session failed for segment with id '{}' and repair number '{}'", segmentId, repairNumber);
+        LOG.warn(
+            "repair session failed for segment with id '{}' and repair number '{}'",
+            segmentId,
+            repairNumber);
         failOutsideSynchronizedBlock = true;
+        // Since we can get out of order notifications,
+        // we need to exit if we already got the COMPLETE notification.
```
Comment (michaelsembwever) [inline src/server/src/main/java/io/cassandrareaper/service/SegmentRunner.java:967]: can we add an `assert` here that success hasn't already been marked true.

File: src/server/src/main/java/io/cassandrareaper/service/SegmentRunner.java [hunk 9]
```diff
@@ -894,6 +920,14 @@ private boolean handleJmxNotificationForCassandra21(
                     .withId(segmentId)
                     .build());
 
+            // Since we can get out of order notifications,
+            // we need to exit if we already got the COMPLETE notification.
```
Comment (michaelsembwever) [inline src/server/src/main/java/io/cassandrareaper/service/SegmentRunner.java:940]: can we add an `assert` here that failed hasn't already been marked true.

File: src/server/src/main/java/io/cassandrareaper/service/SegmentRunner.java [hunk 10]
```diff
@@ -823,17 +837,29 @@ private boolean handleJmxNotificationForCassandra22(
             segmentId,
             repairNumber);
         failOutsideSynchronizedBlock = true;
```
Comment (michaelsembwever) [inline src/server/src/main/java/io/cassandrareaper/service/SegmentRunner.java:845]: can we add an `assert` here that success hasn't already been marked true.

File: src/server/src/main/java/io/cassandrareaper/service/SegmentRunner.java [hunk 11]
```diff
@@ -808,6 +814,14 @@ private boolean handleJmxNotificationForCassandra22(
                     .withEndTime(DateTime.now())
                     .withId(segmentId)
                     .build());
+
```
Comment (michaelsembwever) [inline src/server/src/main/java/io/cassandrareaper/service/SegmentRunner.java:818]: can we add an `assert` here that failed hasn't already been marked true.

**Review by michaelsembwever — CHANGES_REQUESTED** [review]

File: src/server/src/test/java/io/cassandrareaper/service/SegmentRunnerTest.java (hunk 1 above)
Comment (michaelsembwever) [reply src/server/src/test/java/io/cassandrareaper/service/SegmentRunnerTest.java]: can we get a variations of this test for 
 - Cassandra-2.1 and Cassandra-2.2
 - all different sequences of notifications (out of order `failed` isn't tested)

**Review by michaelsembwever — CHANGES_REQUESTED** [review]

File: src/server/src/main/java/io/cassandrareaper/service/SegmentRunner.java (hunk 4 above)
Comment (adejanovski) [reply src/server/src/main/java/io/cassandrareaper/service/SegmentRunner.java]: Just a habit I've taken... I'll switch back to `true`.

File: src/server/src/main/java/io/cassandrareaper/service/SegmentRunner.java (hunk 2 above)
Comment (adejanovski) [reply src/server/src/main/java/io/cassandrareaper/service/SegmentRunner.java]: ok, but I was just being consistent with the way we deal with `segmentFailed` ;)

File: src/server/src/main/java/io/cassandrareaper/service/SegmentRunner.java (hunk 3 above)
Comment (adejanovski) [reply src/server/src/main/java/io/cassandrareaper/service/SegmentRunner.java]: sure thing

File: src/server/src/test/java/io/cassandrareaper/service/SegmentRunnerTest.java (hunk 1 above)
Comment (adejanovski) [reply src/server/src/test/java/io/cassandrareaper/service/SegmentRunnerTest.java]: cool

File: src/server/src/main/java/io/cassandrareaper/service/RepairRunner.java (hunk 5 above)
Comment (adejanovski) [reply src/server/src/main/java/io/cassandrareaper/service/RepairRunner.java:156]: Not really concerned with performance here, I'd rather avoid adding lines and conditional blocks if you don't mind.

File: src/server/src/main/java/io/cassandrareaper/service/SegmentRunner.java (hunk 6 above)
Comment (adejanovski) [reply src/server/src/main/java/io/cassandrareaper/service/SegmentRunner.java]: This is precisely what we're trying to deal with. We get cases where `COMPLETE` gets received before `SUCCESS` does.
What makes you think it can never be true ?

File: src/server/src/main/java/io/cassandrareaper/service/SegmentRunner.java (hunk 7 above)
Comment (adejanovski) [reply src/server/src/main/java/io/cassandrareaper/service/SegmentRunner.java]: same here

File: src/server/src/main/java/io/cassandrareaper/service/SegmentRunner.java (hunk 9 above)
Comment (adejanovski) [reply src/server/src/main/java/io/cassandrareaper/service/SegmentRunner.java:940]: `failedNotified` or `segmentFailed` ?

In any case, what's the point ? There would be a big fat bug in Cassandra if we could get a `SESSION_FAILED` notif and then a `SESSION_SUCCESS`. 
It shouldn't be up to Reaper to assert here, as it's not anymore a matter of race conditions (which Reaper can/should handle).

File: src/server/src/main/java/io/cassandrareaper/service/SegmentRunner.java (hunk 10 above)
Comment (adejanovski) [reply src/server/src/main/java/io/cassandrareaper/service/SegmentRunner.java:845]: Same as before, I don't see the point.

File: src/server/src/main/java/io/cassandrareaper/service/SegmentRunner.java (hunk 11 above)
Comment (adejanovski) [reply src/server/src/main/java/io/cassandrareaper/service/SegmentRunner.java:818]: again

**Comment by adejanovski** [comment]
I'll rebase, address some of your comments and add a test for out of order failed notif (and duplicate the tests for 2.1/2.2).

File: src/server/src/main/java/io/cassandrareaper/service/SegmentRunner.java (hunk 9 above)
Comment (michaelsembwever) [reply src/server/src/main/java/io/cassandrareaper/service/SegmentRunner.java:940]: that's the point of assertions, to assert something we presume is true.
it also serves as documentation (another contributor to reaper may not know that about Cassandra)
(it also good practice to be defensive against external systems)
(it is a use-case where i'd use `assert ` instead of `Preconditions.check…()`)

File: src/server/src/main/java/io/cassandrareaper/service/SegmentRunner.java [hunk 12]
```diff
@@ -96,6 +96,9 @@
   private int commandId;
   private final AtomicBoolean segmentFailed;
   private final UUID leaderElectionId;
+  private final AtomicBoolean failedNotified = new AtomicBoolean(false);
+  private final AtomicBoolean successNotified = new AtomicBoolean(false);
```
Comment (michaelsembwever) [inline src/server/src/main/java/io/cassandrareaper/service/SegmentRunner.java]: thoughts about replacing these two above with just `failedOrSuccessNotified` ?

**Review by michaelsembwever — DISMISSED** [review]
Still keen to see those assertions in place.

File: src/server/src/main/java/io/cassandrareaper/service/SegmentRunner.java (hunk 12 above)
Comment (adejanovski) [reply src/server/src/main/java/io/cassandrareaper/service/SegmentRunner.java]: I made this change.

**Comment by adejanovski** [comment]
Still don't see the value of adding more assertions here as we're not checking something that is generated by Reaper's behavior. If we go down that path we'll end up asserting everything that shouldn't be happening.
Let's move on please ?

**Comment by michaelsembwever** [comment]
> Still don't see the value of adding more assertions here as we're not checking something that is generated by Reaper's behaviour.

I see the value, and it's only four lines of code being requested.

Apart from existing reasons given, the code is introducing an adhoc state-machine with the use of multiple `AtomicBoolean`, and i feel it makes sense that we validate the transitions. 
Doing so also documents the state-machine, as has been said already.

> If we go down that path we'll end up asserting everything that shouldn't be happening.

Adding assertions is a good thing, it adds both defensiveness and documentation, both of which are valuable in open-source code where anyone can (and is encouraged to) contribute.
So yes, if the code is ambiguous (or can permit invalid state in any way) then an assertion/precondition should be used.

File: src/server/src/main/java/io/cassandrareaper/service/SegmentRunner.java [hunk 13]
```diff
@@ -791,6 +791,11 @@ private boolean handleJmxNotificationForCassandra22(
         break;
 
       case SUCCESS:
+
+        Preconditions.checkState(
+            !successOrFailedNotified.get(),
+            "multiple notifications of either 'SUCCESS' or 'FAILURE' are not valid");
+
```
Comment (michaelsembwever) [inline src/server/src/main/java/io/cassandrareaper/service/SegmentRunner.java]: I would like to update (force push) this commit so that these lines contain the repairId and segmentId (so it's not so cryptic to debug in the event it does happen).
(The proposal is in https://github.com/thelastpickle/cassandra-reaper/commit/1cd25771dda94ffc7dfc908d215a4bd6b5df701d )

@adejanovski am i free to force push that commit to this branch?

Comment (adejanovski) [reply src/server/src/main/java/io/cassandrareaper/service/SegmentRunner.java]: @michaelsembwever yes feel free to force push.

Comment (michaelsembwever) [reply src/server/src/main/java/io/cassandrareaper/service/SegmentRunner.java]: thanks. done.

**Review by michaelsembwever — APPROVED** [review]
